### PR TITLE
Increase code coverage for UserProfileService for AB#16696

### PR DESCRIPTION
**Pull request by BrianMaki**
# Implements AB#16996

## Description

Refactored User Profile Service unit tests:

1. Created ApplicationSettingsService

- GetLatestTourChangeDateTimeAsync (formerly UserProfielService)

2. Created LegalAgreementService

- GetActiveTermsOfServiceAsync  (formerly UserProfielService)
- GetActiveLegalAgreementId  (formerly UserProfielService).  Was previously private but now public.

3. Updated UserProfileController to use new services.  This has resulted in one to many constructor parameters by 1.  I put an ignore as was done in UserProfileService.  I believe this is still better than moving some services back into UserProfileService. The services created above are light/flexible and can be easily unit tested. more so than having to do it via UserProfileService tests.

4. Updated UserProfileService to access newly created services
5. Updated UserProfileServiceTests to get 100% code coverage
6. Deleted no longer used mock package and unit test json file
7. Fixed User Feedback Test to get 100% coverage
8. Fixed Rider warnings
9. Fixed sonar code smells
10. Fixed bug in UserProfileService.RecoverUserProfileAsync 'by deleting 'return RequestResultFactory.Success(isValid)' and added 'return' to await this.HandleUpdateUserProfileResultAsync(updateResult, EmailTemplateName.AccountRecoveredTemplate, ct)

**Cypress tests passed with known errors to Jennifer Testfour:**

https://cloud.cypress.io/projects/ofnepc/runs/3922/overview?roarHideRunsWithDiffGroupsAndTags=1

**Sonar:**

https://sonarcloud.io/code?id=bcgov_healthgateway&selected=bcgov_healthgateway%3AApps%2FGatewayApi

## Testing

- [x] Unit Tests Updated
- [ ] Functional Tests Updated
- [ ] Not Required

## Items to Review:

-   [General PR Guidelines](https://github.com/bcgov/healthgateway/wiki/PRguidance)

### CODE REVIEW

File: Apps/GatewayApi/test/unit/Services.Test/UserProfileServiceTests.cs [hunk 1]
```diff
@@ -17,726 +17,1235 @@ namespace HealthGateway.GatewayApiTests.Services.Test
...
+        /// <param name="patientErrorExists">The value indicating whether patient error exists or not.</param>
         /// <returns>A <see cref="Task"/> representing the asynchronous unit test.</returns>
+        [InlineData(0, 19, false)]
+        [InlineData(18, 19, false)]
+        [InlineData(19, 19, true)]
         [Theory]
-        [InlineData(DbStatusCode.Created)]
-        [InlineData(DbStatusCode.Error)]
-        public async Task ShouldCreateUserPreference(DbStatusCode dBStatusCode)
+        public async Task ShouldNotValidateInvalidMinimumAgeAsync(int age, int minAge, bool patientErrorExists)
```
Comment (MikeLyttle) [inline Apps/GatewayApi/test/unit/Services.Test/UserProfileServiceTests.cs:342]: Similar naming issue here - validation implies both accepting valid inputs and rejecting invalid ones, so the test shouldn't be to "not validate".

Again, I'd suggest combining with the previous test - `ShouldValidateMinimumAgeAsync` - or renaming both:

- `ShouldValidateValidMinimumAge` and `ShouldValidateInvalidMinimumAge`, or
- `ShouldValidateMinimumAgeSuccess` and `ShouldValidateMinimumAgeFailure`, or
- `ShouldReturnSuccessForValidateMinimumAgeAsync` and `ShouldReturnFailureForValidateMinimumAgeAsync`, or
- `ValidateMinimumAgeAsyncShouldReturnSuccess` and `ValidateMinimumAgeAsyncShouldReturnFailure`

File: Apps/GatewayApi/test/unit/Services.Test/ReportServiceTests.cs [hunk 2]
```diff
@@ -79,5 +87,32 @@ public async Task ShouldGetReport(TemplateType templateType, string reportName)
...
+            // Arrange
+            ReportRequestModel reportRequest = new()
+            {
+                Data = JsonDocument.Parse("{}").RootElement,
+                Template = TemplateType.Covid,
+                Type = (ReportFormatType)999,
+            };
+
+            string resourceName = $"HealthGateway.GatewayApi.Assets.Templates.{reportRequest.Template}Report.";
+            string expected = $"Template {resourceName} not found.";
```
[the remark this answers is not in the document]
Comment (BrianMaki) [reply Apps/GatewayApi/test/unit/Services.Test/ReportServiceTests.cs:107]: Merged resourceName into expected.  Added comment on line 103.

File: Apps/GatewayApi/test/unit/Services.Test/UserProfileServiceTests.cs [hunk 3]
```diff
@@ -17,726 +17,1235 @@ namespace HealthGateway.GatewayApiTests.Services.Test
...
+        /// IsPhoneNumberValidAsync.
         /// </summary>
+        /// <param name="phoneNumber">The phone number to validate.</param>
         /// <returns>A <see cref="Task"/> representing the asynchronous unit test.</returns>
-        [Fact]
-        public async Task ShouldQueueNotificationUpdate()
+        [InlineData("3345678901")]
+        [InlineData("2507001000")]
+        [Theory]
+        public async Task ShouldPhoneNumberBeValidAsync(string phoneNumber)
```
[the remark this answers is not in the document]
Comment (BrianMaki) [reply Apps/GatewayApi/test/unit/Services.Test/UserProfileServiceTests.cs:276]: Changed names of both tests as per comment.

File: Apps/GatewayApi/test/unit/Services.Test/UserProfileServiceTests.cs (hunk 1 above)
Comment (BrianMaki) [reply Apps/GatewayApi/test/unit/Services.Test/UserProfileServiceTests.cs:342]: Changed names to ShouldValidateValidMinimumAge and ShouldValidateInvalidMinimumAge

File: Apps/GatewayApi/test/unit/Services.Test/ApplicationSettingsServiceTests.cs [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+                            It.Is<string>(key => key.Contains(cacheKey)),
+                            It.IsAny<Func<Task<DateTime?>>>(),
+                            It.IsAny<TimeSpan?>(),
+                            It.IsAny<CancellationToken>()))
+                    .ReturnsAsync(
+                        (string _, Func<Task<DateTime?>> valueFactory, TimeSpan _, CancellationToken _) =>
+                        {
+                            Task<DateTime?> value = valueFactory.Invoke();
+                            return value.Result;
+                        });
```
[the remark this answers is not in the document]
Comment (BrianMaki) [reply Apps/GatewayApi/test/unit/Services.Test/ApplicationSettingsServiceTests.cs:102]: Used option 2.

**Review by MikeLyttle — APPROVED** [review]

**Pull request merged**